// @portabletext/editor/lib/selectors/index.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: !0 });
var selectors = require("../_chunks-cjs/selectors.cjs"), selector_getTextBefore = require("../_chunks-cjs/selector.get-text-before.cjs");
const getActiveListItem = ({
  context
}) => {
  if (!context.selection)
    return;
  const guards = selectors.createGuards(context), selectedTextBlocks = selectors.getSelectedBlocks({
    context
  }).map((block) => block.node).filter(guards.isTextBlock), firstTextBlock = selectedTextBlocks.at(0);
  if (!firstTextBlock)
    return;
  const firstListItem = firstTextBlock.listItem;
  if (firstListItem && selectedTextBlocks.every((block) => block.listItem === firstListItem))
    return firstListItem;
};
exports.getFirstBlock = selectors.getFirstBlock;
exports.getFocusBlock = selectors.getFocusBlock;
exports.getFocusBlockObject = selectors.getFocusBlockObject;
exports.getFocusChild = selectors.getFocusChild;
exports.getFocusListBlock = selectors.getFocusListBlock;
exports.getFocusSpan = selectors.getFocusSpan;
exports.getFocusTextBlock = selectors.getFocusTextBlock;
exports.getLastBlock = selectors.getLastBlock;
exports.getNextBlock = selectors.getNextBlock;
exports.getPreviousBlock = selectors.getPreviousBlock;
exports.getSelectedBlocks = selectors.getSelectedBlocks;
exports.getSelectionEndBlock = selectors.getSelectionEndBlock;
exports.getSelectionStartBlock = selectors.getSelectionStartBlock;
exports.selectionIsCollapsed = selectors.selectionIsCollapsed;
exports.getBlockTextBefore = selector_getTextBefore.getBlockTextBefore;
exports.getSelectionText = selector_getTextBefore.getSelectionText;
exports.getActiveListItem = getActiveListItem;
