// @portabletext/editor/lib/selectors/index.js
import { getSelectedBlocks, createGuards } from "../_chunks-es/selectors.js";
import { getFirstBlock, getFocusBlock, getFocusBlockObject, getFocusChild, getFocusListBlock, getFocusSpan, getFocusTextBlock, getLastBlock, getNextBlock, getPreviousBlock, getSelectionEndBlock, getSelectionStartBlock, selectionIsCollapsed } from "../_chunks-es/selectors.js";
import { getBlockTextBefore, getSelectionText } from "../_chunks-es/selector.get-text-before.js";
const getActiveListItem = ({
  context
}) => {
  if (!context.selection)
    return;
  const guards = createGuards(context), selectedTextBlocks = getSelectedBlocks({
    context
  }).map((block) => block.node).filter(guards.isTextBlock), firstTextBlock = selectedTextBlocks.at(0);
  if (!firstTextBlock)
    return;
  const firstListItem = firstTextBlock.listItem;
  if (firstListItem && selectedTextBlocks.every((block) => block.listItem === firstListItem))
    return firstListItem;
};
export {
  getActiveListItem,
  getBlockTextBefore,
  getFirstBlock,
  getFocusBlock,
  getFocusBlockObject,
  getFocusChild,
  getFocusListBlock,
  getFocusSpan,
  getFocusTextBlock,
  getLastBlock,
  getNextBlock,
  getPreviousBlock,
  getSelectedBlocks,
  getSelectionEndBlock,
  getSelectionStartBlock,
  getSelectionText,
  selectionIsCollapsed
};
